Play computer

dhirallin opened the pull request:
Phew that took a while. Finally finished implementing the "Play Computer" option.
It reuses some code by sending messages to GameMove: and GameEnd: i.e. simulating messages from the server, except adjusted to stay persistent when offline or losing/gaining connection

In terms of difficulty settings, it's currently using the following:

Stockfish Skill Level for each Freechess level (1-10) 
[0, 1, 2, 3, 5, 7, 9, 11, 13, 15]
So I try not to escelate the difficulty too quickly
Even though Stockfish goes upto level 20. I made the max 15, because I'm not sure there's much point in making it impossible to beat. This may need some adjusting though.
Also using maximum movetime (ms) of
[100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]
so it thinks 100ms longer per move for each difficulty level upto 1s
No real logic for that, just wanted to make sure it thinks long enough at the harder difficulties.
Currently it's not limiting the search depth, because I'm not sure it's necessary to make it even easier. 
But tell me what you think, whether it seems too hard at the beginning levels?
Also it turns out that the version of Stockfish we are using supports crazyhouse, so I added that into the engine. 

Another thing, there is currently no opening book, so the openings might get a bit repetitive.
Stockfish apparently doesn't support opening books anymore. So to implement that, we would have to use an npm package like chess-polyglot or chess-tools to load in a book file. Then probably choose a random book line and use some kind of cool down function to randomely deviate from the book. 

Secondly I implemented the suggested idea for the clock. Now when it gets below 20 seconds remaining it shows a 10ths of a second digit counting down. This digit is displayed in a smaller font to the main clock. 

TODO:
I think there still needs to be some cosmetic changes to the top and bottom panels. On mobile when the player's name is long or there's lots of captured pieces, the items are wrapping and it looks really awful. So probably need to truncate the names, or make the fonts smaller, or even overlay the captured pieces and clock over the top of the player name etc. 

Also it'd be good if there was an indicator for whose turn it is, like change the background color behind the clock etc. At the momeent the only indication of whose turn it is comes from the square highlighting on the board. 

I'd also like to add more buttons to the move list, such as the ability to view the move list in pgn form, which would probably be better suited for mobile, and add some load/save pgn functions and load fen/copy fen to clipboard etc etc.

freechessclub-dev commented:
Amazing. I looked through the changes and they LGTM. Thanks for taking this one on!

> It reuses some code by sending messages to GameMove: and GameEnd: i.e. simulating messages from the server,

That's nifty. This is how I was hoping we'd do it :)

> In terms of difficulty settings, it's currently using the following:

The difficulty logic looks reasonable to me. I tried it out at the easy and medium levels and it seemed fine. I'll try some more 
later and will have more to say then. We can fine-tune the logic if it seems too hard.

> Another thing, there is currently no opening book, so the openings might get a bit repetitive.
Stockfish apparently doesn't support opening books anymore. So to implement that, we would have to use an npm package like chess-polyglot or chess-tools to load in a book file. Then probably choose a random book line and use some kind of cool down function to randomely deviate from the book.

Yeah, sounds easy enough that we could consider adding it in the future. Let me give it a whirl and see how repetitive it really feels.

+1 to all the TODOs. I'm going to go ahead, merge this PR, prepare a 3.0.0 release, and deploy it now!

freechessclub-dev commented:
BTW, on difficulty: I realized that skill can go negative too. Setting depth and negative skill may probably be better for the easier levels. Let me play some more though and get a better sense for it.

lanternbugs commented:
> Phew that took a while. Finally finished implementing the "Play Computer" option. It reuses some code by sending messages to GameMove: and GameEnd: i.e. simulating messages from the server, except adjusted to stay persistent when offline or losing/gaining connection
> 
> In terms of difficulty settings, it's currently using the following:
> 
> Stockfish Skill Level for each Freechess level (1-10) [0, 1, 2, 3, 5, 7, 9, 11, 13, 15] So I try not to escelate the difficulty too quickly Even though Stockfish goes upto level 20. I made the max 15, because I'm not sure there's much point in making it impossible to beat. This may need some adjusting though. Also using maximum movetime (ms) of [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000] so it thinks 100ms longer per move for each difficulty level upto 1s No real logic for that, just wanted to make sure it thinks long enough at the harder difficulties. Currently it's not limiting the search depth, because I'm not sure it's necessary to make it even easier. But tell me what you think, whether it seems too hard at the beginning levels? Also it turns out that the version of Stockfish we are using supports crazyhouse, so I added that into the engine.
> 
> Another thing, there is currently no opening book, so the openings might get a bit repetitive. Stockfish apparently doesn't support opening books anymore. So to implement that, we would have to use an npm package like chess-polyglot or chess-tools to load in a book file. Then probably choose a random book line and use some kind of cool down function to randomely deviate from the book.
> 
> Secondly I implemented the suggested idea for the clock. Now when it gets below 20 seconds remaining it shows a 10ths of a second digit counting down. This digit is displayed in a smaller font to the main clock.
> 
> TODO: I think there still needs to be some cosmetic changes to the top and bottom panels. On mobile when the player's name is long or there's lots of captured pieces, the items are wrapping and it looks really awful. So probably need to truncate the names, or make the fonts smaller, or even overlay the captured pieces and clock over the top of the player name etc.
> 
> Also it'd be good if there was an indicator for whose turn it is, like change the background color behind the clock etc. At the momeent the only indication of whose turn it is comes from the square highlighting on the board.
> 
> I'd also like to add more buttons to the move list, such as the ability to view the move list in pgn form, which would probably be better suited for mobile, and add some load/save pgn functions and load fen/copy fen to clipboard etc etc.

i'm not sure limiting difficulty by search time is the best way to do it. From what i've seen it's unreliable at low times to really know how much cpu time you get because of things like OS switching in the processes running.  I would use the explicit UI for difficulty with the UCI Commands given here https://disservin.github.io/stockfish-docs/stockfish-wiki/UCI-&-Commands.html.  Things like UCI_LimitStrength and UCI_Elo.   straight limiting search depth like this is my depth one level this is my depth two level i feel from bots i've run works better than search time changes if you want a second route.

lanternbugs commented:
i sent adcool a 285k text file of book lines that can be used to build a book of mainstream openings. My pulsar engine builds a book on launch and only starts to call the engine to play when out of book.

lanternbugs commented:
[bigbook.txt](https://github.com/freechessclub/freechessclub-app/files/14518770/bigbook.txt)
guess i can drop it here

dhirallin commented:
> BTW, on difficulty: I realized that skill can go negative too. Setting depth and negative skill may probably be better for the easier levels. Let me play some more though and get a better sense for it.

I tried this, but I was getting strange results. For example I was getting my butt handed to me when I set Skill Level to -9. Perhaps that was for a different version of Stockfish?

dhirallin commented:
> > Phew that took a while. Finally finished implementing the "Play Computer" option. It reuses some code by sending messages to GameMove: and GameEnd: i.e. simulating messages from the server, except adjusted to stay persistent when offline or losing/gaining connection
> > In terms of difficulty settings, it's currently using the following:
> > Stockfish Skill Level for each Freechess level (1-10) [0, 1, 2, 3, 5, 7, 9, 11, 13, 15] So I try not to escelate the difficulty too quickly Even though Stockfish goes upto level 20. I made the max 15, because I'm not sure there's much point in making it impossible to beat. This may need some adjusting though. Also using maximum movetime (ms) of [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000] so it thinks 100ms longer per move for each difficulty level upto 1s No real logic for that, just wanted to make sure it thinks long enough at the harder difficulties. Currently it's not limiting the search depth, because I'm not sure it's necessary to make it even easier. But tell me what you think, whether it seems too hard at the beginning levels? Also it turns out that the version of Stockfish we are using supports crazyhouse, so I added that into the engine.
> > Another thing, there is currently no opening book, so the openings might get a bit repetitive. Stockfish apparently doesn't support opening books anymore. So to implement that, we would have to use an npm package like chess-polyglot or chess-tools to load in a book file. Then probably choose a random book line and use some kind of cool down function to randomely deviate from the book.
> > Secondly I implemented the suggested idea for the clock. Now when it gets below 20 seconds remaining it shows a 10ths of a second digit counting down. This digit is displayed in a smaller font to the main clock.
> > TODO: I think there still needs to be some cosmetic changes to the top and bottom panels. On mobile when the player's name is long or there's lots of captured pieces, the items are wrapping and it looks really awful. So probably need to truncate the names, or make the fonts smaller, or even overlay the captured pieces and clock over the top of the player name etc.
> > Also it'd be good if there was an indicator for whose turn it is, like change the background color behind the clock etc. At the momeent the only indication of whose turn it is comes from the square highlighting on the board.
> > I'd also like to add more buttons to the move list, such as the ability to view the move list in pgn form, which would probably be better suited for mobile, and add some load/save pgn functions and load fen/copy fen to clipboard etc etc.
> 
> i'm not sure limiting difficulty by search time is the best way to do it. From what i've seen it's unreliable at low times to really know how much cpu time you get because of things like OS switching in the processes running. I would use the explicit UI for difficulty with the UCI Commands given here https://disservin.github.io/stockfish-docs/stockfish-wiki/UCI-&-Commands.html. Things like UCI_LimitStrength and UCI_Elo. straight limiting search depth like this is my depth one level this is my depth two level i feel from bots i've run works better than search time changes if you want a second route.



> i'm not sure limiting difficulty by search time is the best way to do it. From what i've seen it's unreliable at low times to really know how much cpu time you get because of things like OS switching in the processes running.

That's true, it could be unreliable. I was mainly limiting movetime on easier difficulties for user convenience, because longer move time is not necessary.  But yeah, perhaps movetime shouldn't be used at all, and the time should be limited through max nodes. Could also limit difficulty through max nodes or max depth if Skill Level is not enough.

dhirallin commented:
>  I'm going to go ahead, merge this PR, prepare a 3.0.0 release, and deploy it now!

I'm just doing some cosmetic fixes now. 
Rather than wrapping elements in the top and bottom panels, if the player name is too long, I'm scaling the player/opponent name font sizes down until they reach a min size, and then using overflow-hidden so the names and badges get truncated until they are width 0, then the captured pieces get truncated until they are width 0. So the clock is never obscured or wrapped.

dhirallin commented:
> i sent adcool a 285k text file of book lines that can be used to build a book of mainstream openings. My pulsar engine builds a book on launch and only starts to call the engine to play when out of book.

I think another option is to use a polyglot book file like this one used in SCID. That way we can use polyglot-chess package to navigate it easily.
https://sourceforge.net/p/scidvspc/code/HEAD/tree/books/gm2600.bin
